source/m2mtlvserializer.cpp

```diff
@@ -164,11 +171,61 @@ bool M2MTLVSerializer::serialize_multiple_resource(const M2MResource *resource,
     return success;
 }
 
+/* See, OMA-TS-LightweightM2M-V1_0-20170208-A, Appendix C,
+ * Data Types, Integer, TLV Format 
+ * 
+ * This is easy to change to support 8, 16 or 32 interger. */
+template <typename T> void M2MTLVSerializer::set_value_int(T value, uint8_t *buffer, uint32_t &size)
+{
+    size = sizeof(value);
+
+    /* write bytes to big endian order in buffer */
+    for (uint8_t ind=0;ind<size;ind++) {
+        buffer[size-ind-1] = ((value >> (ind*8)) & 0xFF);
+    }
+}
+
+template void M2MTLVSerializer::set_value_int<int8_t>(int8_t value, uint8_t *buffer, uint32_t &size);
+template void M2MTLVSerializer::set_value_int<int16_t>(int16_t value, uint8_t *buffer, uint32_t &size);
+template void M2MTLVSerializer::set_value_int<int32_t>(int32_t value, uint8_t *buffer, uint32_t &size);
+template void M2MTLVSerializer::set_value_int<int64_t>(int64_t value, uint8_t *buffer, uint32_t &size);
+
 bool M2MTLVSerializer::serialize_resource_instance(uint16_t id, const M2MResourceInstance *resource, uint8_t *&data, uint32_t &size)
 {
-    return serialize_TILV(TYPE_RESOURCE_INSTANCE, id, resource->value(), resource->value_length(), data, size);
+    bool success;
+
+    if ( (resource->resource_instance_type() == M2MResourceInstance::INTEGER) ||
+         (resource->resource_instance_type() == M2MResourceInstance::BOOLEAN) ||
+         (resource->resource_instance_type() == M2MResourceInstance::TIME) ) {
+        success=serialize_TLV_binary_int(resource, TYPE_RESOURCE_INSTANCE, id, data, size);
+        }
+    else {
+        success=serialize_TILV(TYPE_RESOURCE_INSTANCE, id, resource->value(), resource->value_length(), data, size);
+    }
+
+    return success;
+}
+
+template <typename T> bool M2MTLVSerializer::serialize_TLV_binary_int(T *resource, uint8_t type, uint16_t id, uint8_t *&data, uint32_t &size)
+{
+        int64_t valueInt = resource->get_value_int();
+        uint32_t buffer_size;
+        /* max len 8 bytes */
+        uint8_t buffer[8];
+
+        if (resource->resource_instance_type() == M2MResourceInstance::BOOLEAN) {
+            set_value_int((int8_t)valueInt, buffer, buffer_size);
+        }
+        else {
+            set_value_int(valueInt, buffer, buffer_size);
+        }
+
+        return serialize_TILV(type, id, buffer, buffer_size, data, size);
 }
 
+template bool M2MTLVSerializer::serialize_TLV_binary_int<const M2MResourceInstance>(const M2MResourceInstance *resource, uint8_t type, uint16_t id, uint8_t *&data, uint32_t &size);
+template bool M2MTLVSerializer::serialize_TLV_binary_int<const M2MResource>(const M2MResource *resource, uint8_t type, uint16_t id, uint8_t *&data, uint32_t &size);
```

Comment:
Does this even need to be templated? Isn't M2MResourceInstance version enough since M2MResource inherits from that?